man/wall.1
'\" -*- coding: UTF-8 -*-
.\" Copyright (C) 1998-2003 Miquel van Smoorenburg.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
.\"
.TH WALL 1 "15 April 2003" "sysvinit " "User Commands"

.SH NAME
wall - send a message to everybody's terminal

.SH SYNOPSIS
.B wall
.RB [ \-n ]
.RI [ message ]

.SH DESCRIPTION
.B Wall
sends a message to everybody logged in with their
.BR mesg (1)
permission
set to
.BR yes .
The message can be given as an argument to
.BR wall ,
or it can be sent to
.BR wall 's
standard input.  When using the standard input from a terminal,
the message should be terminated with the
.B EOF
key (usually Control-D).
.PP
The length of the message is limited to 20 lines.
For every invocation of
.B wall
a notification will be written to syslog, with facility
.B LOG_USER
and level
.BR LOG_INFO .

.SH OPTIONS
.IP \fB\-n\fP
Suppresses the normal banner printed by
.BR wall ,
changing it to "Remote broadcast message".
This option is only available for root if
.B wall
is installed set-group-id, and is used by
.BR rpc.walld (8).
.PP

.SH ENVIRONMENT
.B wall
ignores the
.B TZ
variable - the time printed in the banner is based on the system's
local time.

.SH SEE ALSO
.BR mesg (1),
.BR rpc.rwalld (8)

.SH AUTHOR
.MT miquels@\:cistron\:.nl
Miquel van Smoorenburg
.ME


doc/Changelog
sysvinit (3.02) unreleased; urgency=low
    * Added q and Q flags to synopsis in shutdown manual page.
    * Applied fixes for markup and spacing in manual pages.
      Patch provided by Mario Blattermann.
    * Added translation framework (po4a) from Mario Blttermann.
    * Added Makefile for man/ directory. Will handle translations
      and substitutions.
    * Applied new translations for multiple languages from Mario Blattermann.
    * Added ability to use "@" symbol in command named in the inittab file. This
      treats commands as literal and does not launch a shell to interpret them.
    * Updated inittab manual page to include overview of symbols which trigger
      a shell interpretor and how to disable them using the @ symbol.


sysvinit (3.01) released; urgency=low
    * Default to showing processes in the uninterruptable state (D).
    The -z flag no longer affects whether processes in D state are shown.
    The -z flag does still toggle whether zombie (Z) processes are shown.

    * Removed unnecessary check which is always true from init tab parsing.

sysvinit (3.00) released; urgency=low

    * Applied patch from Matthias Schiffer which allows bootlogd to read from
      a wider range of consoles. The console name is already passed in from the
      kernel command line using "console=". We no longer filter out names as strictly
      but do now check to confirm the "console=" device points to a valid TTY.

sysvinit (2.99) released; urgency=low

    * Fixed typos and missing underlines in shutdown manual page.
      Corrections provided by Helge Kreutzmann.

sysvinit (2.98) released; urgency=low

    * Fixed time parsing in shutdown when there is a + in front of a 0 time offset.
      Commands with a positive time offset (+1) would work but +0 fails.
      This has been corrected by Arkadiusz Miskiewicz.

sysvinit (2.97) released; urgency=low

    * Check $(ROOT) filesystem for libcrypt instead of a hardcoded path to /usr.
      Added logsave and readbootlog to list of files git ignores.
      - Patches provided by Petr Ovtchenkov.

    * Code clean-up and making sure we avoid freeing unused memory.
      Patch provided by David Hoyer.

    * Added shell script from Trek which converts systemd unit files
      into init.d style scripts.

    * Added patch from Didier Gaudin which allows init to load configuration
      data from files stored in /etc/inittab.d/

    * Added patch from William Shipley which allows shutdown time to be specified
      in the format +hh:mm. This is in addition to the existing formats such as
      hh:mm, +m, and "now".

    * Fixed typos in manual pages. Submitted by Helge Kreutzmann.


sysvinit (2.96) released; urgency=low

  [ Jesse Smith ]

    * Added -z command line parameter to pidof which tells pidof to
      try to find processes in uninterruptible (D) or zombie (Z) states.
      This can cause pidof to hang, but produces a more complete process
      list.
      Closes Savannah bug #56534

    * Reformatted init code to make if/while logic more clear.

    * Cleaned up some output from readbootlog.

    * Added -e flag to bootlogd. When -e is used, data saved
      to the boot log file does not have escape characters
      removed. This means colour and cursor movement codes
      stay in the log file. The may then look nicer when
      read with "less -R', but may appear cluttered or
      out of alignment when viewed with other, plain-text tools.
      When -e is not used, escape characters are removed/filtered.
      Closes Debian bug #672361.

    * Make sure src/Makefile cleans up all executable files
      when parent Makefile calls "make clean".


sysvinit (2.95) released; urgency=low

  [ Jesse Smith ]

    * Fixed various typos and control codes in manual pages.
      Patch provided by Bjarni Ingi Gislason.
    * Dropped "-f" format flag for pidof command as it
      could be used to print information from memory or crash
      pidof. Replaced flag with "-d" which allows for a custom
      separator to be used between listed PIDs.
      Patch supplied by KatolaZ.
    * Updated manual page to describe -h and its modifiers (-H and -P)
      in more detail. Should close Debian bug #374039.
    * Use millisecond delays in init so that shutdown
      can happen without a near-full-second delay after
      all processes have terminated. Replaced do_sleep()
      with do_msleep(), provided by Serge Belyshev.
    * Replaced hardwired sleep constants in init.c with
      defined constants for easy updating in the future.
    * Accepted patch from Luc Gommans which explains why zombie and
      deep sleep processes (Z and D) may be hidden from pidof.
    * Removed link instruction against sepol library from src/Makefile.
      This has no change to the resulting binary, just makes the
      Makefile cleaner.

sysvinit (2.94) released; urgency=low

  [ Jesse Smith ]

    * When the halt command is called with the -p flag (or as poweroff)
      the command now passes the "-h -P" flags to shutdown. This
      in turn sets the INIT_HALT environment variable to POWEROFF.
      Assuming this value is checked by initscripts during the
      shutting down procedure, it should cause the system to
      be powered off.
      If halt is called without -p then the value of INIT_HALT
      is not set and the default action (often set in /etc/default/halt)
      is taken.
    * Removed unnecessary malloc.h includes. Memory allocation
      and freeing is now handled in stdlib.h
    * Added defines for FreeBSD to make some components compile
      on FreeBSD 11.
    * Increased the size of the kernel command line buffer in bootlogd
      from 256 characters to 4096. This size is defined in KERNEL_COMMAND_LENGTH
      for easy modification downstream.
    * Added logsave.c and logsave.8 manual page from e2fsprogs to make
      sure logsave is available to initscripts.
    * Updated src/Makefile to make sure bootlogd compiles with Clang.

src/wall.c
/*
 * wall.c	Write to all users logged in.
 *
 * Usage:	wall [text]
 *
 * Version:	@(#)wall  2.79  12-Sep-2000  miquels@cistron.nl
 *
 *		This file is part of the sysvinit suite,
 *		Copyright (C) 1991-2000 Miquel van Smoorenburg.
 *
 *		This program is free software; you can redistribute it and/or modify
 *		it under the terms of the GNU General Public License as published by
 *		the Free Software Foundation; either version 2 of the License, or
 *		(at your option) any later version.
 *
 *		This program is distributed in the hope that it will be useful,
 *		but WITHOUT ANY WARRANTY; without even the implied warranty of
 *		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *		GNU General Public License for more details.
 *
 *		You should have received a copy of the GNU General Public License
 *		along with this program; if not, write to the Free Software
 *		Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <pwd.h>
#include <syslog.h>
#include <sys/types.h>
#include "init.h"


#define MAXLEN 4096
#define LINE_SIZE 80
#define MAXLINES 20

int main(int argc, char **argv)
{
  char buf[MAXLEN];
  char line[LINE_SIZE + 3];   /* leave room for \r\n\0 */
  int i, f, ch;
  int len = 0;
  int remote = 0;
  char *p;
  char *whoami;
  struct passwd *pwd;

  buf[0] = 0;
  if ((pwd = getpwuid(getuid())) == NULL) {
	if (getuid() == 0)
		whoami = "root";
	else {
		fprintf(stderr, "You don't exist. Go away.\n");
		exit(1);
	}
  } else
	whoami = pwd->pw_name;

  while((ch = getopt(argc, argv, "n")) != EOF)
	switch(ch) {
		case 'n':
			/*
			 *	Undocumented option for suppressing
			 *	banner from rpc.rwalld. Only works if
			 *	we are root or if we're NOT setgid.
			 */
			if (geteuid() != 0 && getgid() != getegid()) {
				fprintf(stderr, "wall -n: not privileged\n");
				exit(1);
			}
			remote = 1;
			break;
		default:
			fprintf(stderr, "usage: wall [message]\n");
			return 1;
			break;
	}

  if ((argc - optind) > 0) {
	for(f = optind; f < argc; f++) {
		len += strlen(argv[f]) + 1;
		if (len >= MAXLEN-4) break;
		strcat(buf, argv[f]);
		if (f < argc-1) strcat(buf, " ");
	}
	strcat(buf, "\r\n");
  } else {
	while(fgets(line, 80, stdin)) {
		/*
		 *	Make sure that line ends in \r\n
		 */
		for(p = line; *p && *p != '\r' && *p != '\n'; p++)
			;
		strcpy(p, "\r\n");
		len += strlen(line);
		if (len >= MAXLEN) break;
		strcat(buf, line);
	}
  }

  i = 0;
  for (p = buf; *p; p++) {
	if (*p == '\n' && ++i >= MAXLINES) {
		*++p = 0;
		break;
	}
  }

  openlog("wall", LOG_PID, LOG_USER);
  syslog(LOG_INFO, "wall: user %s broadcasted %d lines (%zu chars)",
	whoami, i, strlen(buf));
  closelog();

  unsetenv("TZ");
  wall(buf, remote);

  /*NOTREACHED*/
  return 0;
}

